x/gamm/pool-models/balancer/amm.go

```diff
@@ -222,35 +266,62 @@ func (p *Pool) maximalExactRatioJoin(tokensIn sdk.Coins) (numShares sdk.Int, rem
 }
 
 func (p *Pool) JoinPool(_ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, err error) {
+	numShares, newLiquidity, err := p.CalcJoinPoolShares(_ctx, tokensIn, swapFee)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+	p.updateLiquidity(numShares, newLiquidity)
+	return numShares, nil
+}
+
+func (p *Pool) CalcJoinPoolShares(_ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, newLiquidity sdk.Coins, err error) {
 	if tokensIn.Len() == 1 {
-		numShares, err = p.singleAssetJoin(tokensIn[0], swapFee)
-		p.updateLiquidity(numShares, tokensIn)
-		return numShares, err
+		numShares, err = p.calcSingleAssetJoin(tokensIn[0], swapFee)
+		newLiquidity = tokensIn
+		return numShares, newLiquidity, err
 	} else if tokensIn.Len() != p.NumAssets() {
-		return sdk.ZeroInt(), errors.New(
+		return sdk.ZeroInt(), sdk.NewCoins(), errors.New(
 			"balancer pool only supports LP'ing with one asset, or all assets in pool")
 	}
 	// Add all exact coins we can (no swap)
 	numShares, remCoins, err := p.maximalExactRatioJoin(tokensIn)
 	if err != nil {
-		return sdk.ZeroInt(), err
+		return sdk.ZeroInt(), sdk.NewCoins(), err
 	}
-	p.updateLiquidity(numShares, tokensIn.Sub(remCoins))
+	newLiquidity = tokensIn.Sub(remCoins)
 	// if there are coins that couldn't be perfectly joined, do single asset joins for each of them.
 	if !remCoins.Empty() {
 		for _, coin := range remCoins {
-			newShares, err := p.singleAssetJoin(coin, swapFee)
+			newShares, err := p.calcSingleAssetJoin(coin, swapFee)
 			if err != nil {
-				return sdk.ZeroInt(), err
+				return sdk.ZeroInt(), sdk.NewCoins(), err
 			}
-			p.updateLiquidity(newShares, sdk.Coins{coin})
+			newLiquidity = newLiquidity.Add(coin)
```

Comment:
So these single asset joins actual need the liquidity updated in order to work :/
This is because the pool reserves and therefore correct amounts out change at each of these steps.

Because the JoinPool and ExitPool methods have a lot of internal steps pool liquidity updates for correctness, I'd suggest doing this in one of two ways:
* Keeping JoinPool and ExitPool as before, but then make the calc methods do a deep copy of the struct. (Probably easier) I don't know if deepcopy / clone methods are going to be straight forward to make, or we'd have to make a function called "clonePoolAssets" to do it.
* Making a common method, that does all of this logic as before, but with pool assets passed in as an argument rather than read off the pool. (It can mutate this pool assets) Then make the calc method use a cloned pool assets struct, and the join pool use the 'real' pool assets struct.

Reply:
Gotcha! How is this passing tests right now tho :/

Reply:
Uhhhh that's a very good question.

On first thought, maybe it's that we previously didn't support lping at arbitrary ratios, and therefore dont have tests for this? (Which is then concerning/  my bad for not adding tests in the refactoring)